### ISSUE TITLE
Feature Request: implementation of the Muon optimizer

### ISSUE THREAD

**Issue by cdelv** [google-deepmind/optax#1406]
Hi Optax team,

I would like to request the implementation of the Muon optimizer in Optax.

The Muon and MuonClip optimizers have recently been introduced as very fast and efficient optimizers for training deep neural networks. Benchmarks have shown that Muon can significantly accelerate convergence and improve training throughput compared to widely used optimizers like Adam and SGD while maintaining or improving model accuracy. Having Muon available in Optax would be very cool. Its also the optimizer used to train [Kimi 2](https://moonshotai.github.io/Kimi-K2/). 

References:

[Implementation](https://github.com/KellerJordan/Muon)
[Pytorch pull request1](https://github.com/ml-explore/mlx/pull/1914)
[paper](https://arxiv.org/pdf/2502.16982)
[paper showing muon efficiency](https://arxiv.org/pdf/2507.07101)

Would the team consider adding support for this optimizer in a future release? 

Thanks for your consideration!

**Comment by rajasekharporeddy** [comment]
Hi @cdelv,

Thanks for reporting the issue. The `Muon optimizer` is available in `optax` as part of `optax.contrib`. Please have a look at the [documentation](https://optax.readthedocs.io/en/latest/api/contrib.html#muon).

**Comment by rdyro** [comment]
Thanks for reporting this, @rajasekharporeddy is correct, we have an implementation available.

The current version requires muon weights to be 2D, but we're working to lift that requirement!

**Comment by rdyro** [comment]
The changes should now be merged: https://github.com/google-deepmind/optax/pull/1407
